File: Kvasha/756.js
/**
 * @param {string} bottom
 * @param {string[]} allowed
 * @return {boolean}
 */
let pyramidTransition = function (bottom, allowed) {
  let blocks = {};

  for (let item of allowed) {
    let [i, j, k] = item;
    blocks[i + j] ? blocks[i + j].push(k) : (blocks[i + j] = [k]);
  }

  function dfs(row, top, idx) {
    if (row.length === 1) return true;
    if (idx === row.length - 1) return dfs(top, '', 0);

    let [i, j] = [row[idx], row[idx + 1]];
    let possible = blocks[i + j];

    if (!possible) return false;

    for (let block of possible) {
      if (dfs(row, top + block, idx + 1)) return true;
    }

    return false;
  }

  return dfs(bottom, '', 0);
};

File: Рудович/756.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    bool pyramidTransition(string bottom, vector<string> allowed) 
    {
        vector<vector<vector<char>>> all(26, vector<vector<char>>(26));
        for (int i = 0; i < allowed.size(); ++i)
        {
            all[allowed[i][0] - 'A'][allowed[i][1] - 'A'].emplace_back(allowed[i][2]);
        }
        return fun(bottom, all, "");
    }

    bool fun(string bottom, vector<vector<vector<char>>>& all, string str)
    {
        if (bottom.size() == 1)
        {
            return true;
        }
        if (str.size() == bottom.size() - 1)
        {
            return fun(str, all, "");
        }
        for (int i = 0; i < all[bottom[str.size()] - 'A'][bottom[str.size() + 1] - 'A'].size(); ++i)
        {
            if (fun(bottom, all, str + all[bottom[str.size()] - 'A'][bottom[str.size() + 1] - 'A'][i]))
            {
                return true;
            }
        }
        return false;
    } 
};